### PR TITLE
PR: Fix indentation on paste

### LINKED ISSUE #11701: Align indented lines after pasting
Currently, when pasting wholly copied lines onto a new line, the indenting whitespaces are combined as below. In multiline highlighting, this requires to always fully backspace on the line being pasted into (or unindenting), which is tedious and is handled automatically by other IDEs.

image

### PR DESCRIPTION
**Pull request by impact27**
## Description of Changes

depends on https://github.com/jupyter/qtconsole/pull/462

Keep relative indentation between first and following lines. Automatically unindent code if python like.

* [ ] Wrote at least one-line docstrings (for any new functions)
* [ ] Added unit test(s) covering the changes (if testable)

* [ ] Included a screenshot or animation (if affecting the UI, see [Licecap](https://www.cockos.com/licecap/))

### Issue(s) Resolved

Fixes #11701

### Affirmation

By submitting this Pull Request or typing my (user)name below,
I affirm the [Developer Certificate of Origin](https://developercertificate.org)
with respect to all commits and content included in this PR,
and understand I am releasing the same under Spyder's MIT (Expat) license.

I certify the above statement is true and correct:

### CODE REVIEW

**Comment by impact27** [comment]
Rebased on master

File: spyder/utils/clipboard_helper.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+
+class _ClipboardHelp():
+    # Clipboard metadata
+    metadata_hash = None
+    metadata_indent = None
+    metadata_tab_stop_width_spaces = None
+
+
+def get_current_hash():
```
Comment (ccordoba12) [inline spyder/utils/clipboard_helper.py:23]: Why not add the functions in this module to a class? Both the metadata attributes of the class above and these functions could be part of it.

That's unless you want to have a single instance of that class for the entire application (which seems you do).

**Review by ccordoba12 — COMMENTED** [review]
Thanks for this @impact27 and sorry for the big delay reviewing it!

I left some initial comments for you.

File: spyder/plugins/editor/widgets/codeeditor.py [hunk 2]
```diff
@@ -2688,17 +2711,74 @@ def paste(self):
...
+
+        lines_adjustment = clipboard_helper.remaining_lines_adjustment(
+            preceding_text)
+
+        if preceding_text.strip() == "" and first_line.strip != "":
+            # The indent is controlled by preceding_text. Remove extra indent
+            extra_indent = self.get_line_indentation(first_line_selected)
+            lines_adjustment -= extra_indent
+            first_line = self.adjust_indentation(
+                first_line, -extra_indent)
```
[the remark this answers is not in the document]
Comment (impact27) [reply spyder/plugins/editor/widgets/codeeditor.py:2737]: If you select multiline text, it is reasonable to think that the indentation of all lines is linked. If I copy a block of text, it is unlikely that I would want the first line indentation to be modified but not the rest of the lines.
This is the general behaviour in atom for example for any programming language.

File: spyder/plugins/editor/widgets/codeeditor.py [hunk 3]
```diff
@@ -2688,17 +2711,74 @@ def paste(self):
...
+
+        # Fix indentation of multiline text
```
[the remark this answers is not in the document]
Comment (impact27) [reply spyder/plugins/editor/widgets/codeeditor.py:2739]: The preceding_text here is the text between the start of the line and the cursor, not from the start of the file.

File: spyder/plugins/editor/widgets/codeeditor.py [hunk 4]
```diff
@@ -2688,17 +2711,74 @@ def paste(self):
...
         self.sig_text_was_inserted.emit()
 
         self.document_did_change(text)
         self.skip_rstrip = False
 
+    def _save_clipboard_indentation(self):
+        """
+        Save the indentation corresponding to the clipboard data.
+
+        Must be called right after copying.
```
[the remark this answers is not in the document]
Comment (impact27) [reply spyder/plugins/editor/widgets/codeeditor.py:2772]: If you copy a block of text, you know the indentation of every line except for the first line. So you must save the indentation of the first line in order to correctly handle this case. (Usually a user would select a multiline block of text by selecting right before the first word, not at the beginning of the line.)

File: spyder/plugins/ipythonconsole/widgets/shell.py [hunk 5]
```diff
@@ -811,6 +812,106 @@ def request_restart_kernel(self):
...
+
+        if indent_adjustment > 0:
+            return ' ' * indent_adjustment + line
+
+        max_indent = clipboard_helper.get_line_indentation(line)
+        indent_adjustment = min(max_indent, -indent_adjustment)
+
+        return line[indent_adjustment:]
+
+    def paste(self, mode=QtGui.QClipboard.Clipboard):
```
[the remark this answers is not in the document]
Comment (impact27) [reply spyder/plugins/ipythonconsole/widgets/shell.py:828]: I think the pasting should be consistent in the console and in the editor

File: spyder/utils/clipboard_helper.py (hunk 1 above)
Comment (impact27) [reply spyder/utils/clipboard_helper.py:23]: Yes the clipboard should be shared for the whole application as you need to save the indentation of the first line. (So you can paste from the console to the editor for example)

File: spyder/plugins/editor/widgets/codeeditor.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (impact27) [reply spyder/plugins/editor/widgets/codeeditor.py:2737]: See implementation in atom for example: https://github.com/atom/atom/blob/976cb9ef3a611163052f9d31c6c3685dc1e6c5b4/src/selection.js#L501

File: spyder/utils/clipboard_helper.py (hunk 1 above)
Comment (ccordoba12) [reply spyder/utils/clipboard_helper.py:23]: Well, in that case you could create a singleton by instantiating the class I'm proposing for this at the end of the file and using it in all places you need.

That's what we do with `CONF` at the end of `spyder/config/manager.py` and the new `PLUGIN_REGISTRY` at the end of `spyder/api/plugin_registration/registry.py`

**Comment by ccordoba12** [comment]
Besides the last comment I left, this also has a conflict now.

File: spyder/utils/clipboard_helper.py [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+from spyder.py3compat import to_text_string
+
+
+class ClipboardHelp:
+    # Clipboard metadata
+    metadata_hash = None
+    metadata_indent = None
+    metadata_tab_stop_width_spaces = None
+
+
```
Comment (ccordoba12) [inline spyder/utils/clipboard_helper.py:21]:
```suggestion
```

File: spyder/utils/clipboard_helper.py [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+    def get_current_hash(self):
+        clipboard = QApplication.clipboard()
+        return hash(to_text_string(clipboard.text()))
+    
+
```
Comment (ccordoba12) [inline spyder/utils/clipboard_helper.py:26]:
```suggestion
```

File: spyder/utils/clipboard_helper.py [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+        return hash(to_text_string(clipboard.text()))
+    
+    
+    def get_line_indentation(self, text, tab_stop_width_spaces=None):
+        """Get indentation for given line."""
+        if tab_stop_width_spaces:
+            text = text.replace("\t", " " * tab_stop_width_spaces)
+        return len(text)-len(text.lstrip())
+    
+
```
Comment (ccordoba12) [inline spyder/utils/clipboard_helper.py:33]:
```suggestion
```

File: spyder/utils/clipboard_helper.py [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+        Save the indentation corresponding to the clipboard data.
+    
+        Must be called right after copying.
+        """
+        self.metadata_hash = self.get_current_hash()
+        self.metadata_indent = self.get_line_indentation(
+            preceding_text, tab_stop_width_spaces)
+        self.metadata_tab_stop_width_spaces = tab_stop_width_spaces
+    
+
```
Comment (ccordoba12) [inline spyder/utils/clipboard_helper.py:45]:
```suggestion
```

File: spyder/utils/clipboard_helper.py [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+"""
+Clipboard helper module.
+"""
+from qtpy.QtWidgets import QApplication
+
+from spyder.py3compat import to_text_string
+
+
+class ClipboardHelp:
```
Comment (ccordoba12) [inline spyder/utils/clipboard_helper.py:15]: I suggest to rename this to `ClipboardHelper`

File: spyder/utils/clipboard_helper.py [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+        if self.get_current_hash() == self.metadata_hash:
+            return (
+                self.get_line_indentation(
+                    preceding_text,
+                    self.metadata_tab_stop_width_spaces)
+                - self.metadata_indent)
+        return 0
+
+
+CLIPHELP = ClipboardHelp()
```
Comment (ccordoba12) [inline spyder/utils/clipboard_helper.py:60]: I also suggest to rename this instance to `CLIPBOARD_HELPER`

**Review by ccordoba12 — COMMENTED** [review]
@impact27, final review, then this should be ready.

**Review by ccordoba12 — APPROVED** [review]
@impact27, thanks for your help with this!

*Note*: The failures in our tests are unrelated to this and are fixed in PR #16026.

**Pull request merged**